clients/rospy/src/rospy/core.py

```diff
@@ -143,78 +145,96 @@ def rospywarn(msg, *args):
 
-loginfo = logging.getLogger('rosout').info
 logout = loginfo # alias deprecated name
 
-logerr = logging.getLogger('rosout').error
+logdebug = partial(_base_logger, logger_level='debug')
+
+logwarn = partial(_base_logger, logger_level='warn')
+
+logerr = partial(_base_logger, logger_level='error')
+
 logerror = logerr # alias logerr
 
-logfatal = logging.getLogger('rosout').critical
-
-
-class LoggingThrottle(object):
-
-    last_logging_time_table = {}
-
-    def __call__(self, caller_id, logging_func, period, msg):
-        """Do logging specified message periodically.
-
-        - caller_id (str): Id to identify the caller
-        - logging_func (function): Function to do logging.
-        - period (float): Period to do logging in second unit.
-        - msg (object): Message to do logging.
-        """
-        now = rospy.Time.now()
-
-        last_logging_time = self.last_logging_time_table.get(caller_id)
-
-        if (last_logging_time is None or
-              (now - last_logging_time) > rospy.Duration(period)):
-            logging_func(msg)
-            self.last_logging_time_table[caller_id] = now
-
-
-_logging_throttle = LoggingThrottle()
-
-
-def _frame_record_to_caller_id(frame_record):
-    frame, _, lineno, _, code, _ = frame_record
-    caller_id = (
-        inspect.getabsfile(frame),
-        lineno,
-        frame.f_lasti,
-    )
-    return pickle.dumps(caller_id)
-
-
-def logdebug_throttle(period, msg):
-    caller_id = _frame_record_to_caller_id(inspect.stack()[1])
-    _logging_throttle(caller_id, logdebug, period, msg)
-
-
-def loginfo_throttle(period, msg):
-    caller_id = _frame_record_to_caller_id(inspect.stack()[1])
-    _logging_throttle(caller_id, loginfo, period, msg)
-
-
-def logwarn_throttle(period, msg):
-    caller_id = _frame_record_to_caller_id(inspect.stack()[1])
-    _logging_throttle(caller_id, logwarn, period, msg)
-
-
-def logerr_throttle(period, msg):
-    caller_id = _frame_record_to_caller_id(inspect.stack()[1])
-    _logging_throttle(caller_id, logerr, period, msg)
-
-
-def logfatal_throttle(period, msg):
-    caller_id = _frame_record_to_caller_id(inspect.stack()[1])
-    _logging_throttle(caller_id, logfatal, period, msg)
-
+logfatal = partial(_base_logger, logger_level='critical')
+
+
+class LoggingThrottle(object):
+
+    LogEntry = namedtuple("LogEntry", "time digest")
+
+    last_log_entry = {}
+
+    def __call__(self, caller_id, logging_func, period, msg, *args):
+        """Do logging specified message periodically. Messages with different contents will bypass throttling
+
+        - caller_id (str): Id to identify the caller
+        - logging_func (function): Function to do logging.
+        - period (float): Period to do logging in seconds.
+        - msg (object): Message to do logging.
+        """
+        now = rospy.Time.now()
+
+        last = self.last_log_entry.get(caller_id, self.LogEntry(time=None, digest=None))
+        digest = hash(*args + tuple([str(msg)]))

```

Comment:
We can remove that, this throttles just messages that don't change.  I can add this to the PR comments if this is something that adds benefit but if it doesn't I will remove.